raiden/network/proxies/token.py

```diff
@@ -27,6 +32,80 @@
 GAS_REQUIRED_FOR_APPROVE = 58792
 
 
+@dataclass
+class ApproveConditionsData:
+    """ Values used in assert or require expressions in the smart contract. """
+
+    balance: int
+
+
+class ApproveUncheckedTransaction:
+    def __init__(self, allowed_address: Address, allowance: TokenAmount) -> None:
+        self.allowed_address = allowed_address
+        self.allowance = allowance
+
+    def estimate_gas(self, token: "Token") -> "ApprovePendingTransaction":
+        gas_limit = token.proxy.estimate_gas(
+            token.client.get_checking_block(), "approve", self.allowed_address, self.allowance
+        )
+
+        if gas_limit is not None:
+            return ApprovePendingTransaction(self, safe_gas_limit(gas_limit))
+
+        failed_at = token.client.get_block("latest")
+        failed_at_blockhash = encode_hex(failed_at["hash"])
+        failed_at_blocknumber = failed_at["number"]
+
+        token.client.check_for_insufficient_eth(
+            transaction_name="approve",
+            transaction_executed=False,
+            required_gas=GAS_REQUIRED_FOR_APPROVE,
+            block_identifier=failed_at_blocknumber,
+        )
+
+        approve_data = token.approve_data(failed_at_blockhash)
+
+        if approve_data.balance < self.allowance:
+            msg = (
+                f"Call to approve will fail. Your balance of {approve_data.balance} is "
+                "below the required amount of {allowance}."
+            )
+            if approve_data.balance == 0:
+                msg += (
+                    " Note: The balance was 0, which may also happen if the contract "
+                    "is not a valid ERC20 token (balanceOf method missing)."
+                )
+            raise RaidenRecoverableError(msg)
+
+        raise RaidenRecoverableError(
```

Comment:
raising the exceptions is not my preferred design but it introduces less changes to the current proxies (and simplifies a bit the signature of this method)